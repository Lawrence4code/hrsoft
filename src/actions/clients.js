import uuid from 'uuid';
import database from '../firebase/firebase';
import { Children } from 'react';
// action generator for adding client
// passed with default value if no value passed for the client
export const addClient = client => ({
  type: 'ADD_CLIENT',
  client
});

export const startAddClient = (clientData = {}) => {
  console.log('startAddClient triggered');
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      firstName = '',
      lastName = '',
      gender = '',
      phone = '',
      email = '',
      address = '',
      status = '',
      field = '',
      note = '',
      lastCommuniation = '',
      memberType = '',
      createdAt = 0
    } = clientData;

    const client = {
      title,
      firstName,
      lastName,
      gender,
      phone,
      email,
      address,
      status,
      field,
      note,
      lastCommuniation,
      memberType,
      createdAt
    };

    // console.log(client);

    return database
      .ref(`users/${uid}/clients`)
      .push(client)
      .then(ref => {
        // console.log(ref.key);
        dispatch(
          addClient({
            id: ref.key,
            ...client
          })
        );
      });
  };
};

// action generating for removing client
export const removeClient = ({ id } = {}) => {
  return {
    type: 'REMOVE_CLIENT',
    id
  };
};

export const startRemoveClient = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients/${id}`)
      .remove()
      .then(() => {
        dispatch(removeClient({ id }));
      });
  };
};

// action generating for modifying the client details
export const editClient = (id, updates) => {
  return {
    type: 'EDIT_CLIENT',
    id,
    updates
  };
};

export const startEditClient = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editClient(id, updates));
      });
  };
};

// set clients

export const setClients = clients => {
  return {
    type: 'SET_CLIENTS',
    clients
  };
};

// export const startSetClients

export const startSetClients = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients`)

      .once('value')
      .then(snapshot => {
        const clients = [];
        snapshot.forEach(childSnapshot => {
          clients.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setClients(clients));
      });
  };
};
