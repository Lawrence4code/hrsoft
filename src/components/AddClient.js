import React from 'react';
import { Link } from 'react-router-dom';

const AddClient = () => (
  <div>
    <h2>AddClientPage</h2>
    <Link to="/"> Back to Dashboard </Link>
  </div>
);

export default AddClient;
