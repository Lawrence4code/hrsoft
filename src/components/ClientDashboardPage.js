import React from 'react';
import { Link } from 'react-router-dom';
import ClientList from './ClientList';
import ClientListFilters from './ClientListFilters';
import ClientSummary from './ClientSummary';

const ClientDashboardPage = () => (
  <div>
    <ClientSummary />
    <ClientListFilters />
    <ClientList />
  </div>
);

export default ClientDashboardPage;
