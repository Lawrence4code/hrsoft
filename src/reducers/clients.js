// default values for the cient reducer
const clientReducerDefaultState = [];

const clientsReducer = (state = clientReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_CLIENT':
      return [...state, action.client];
    case 'EDIT_CLIENT':
      return state.map(client => {
        if (client.id === action.id) {
          return {
            ...client,
            ...action.updates
          };
        } else {
          return client;
        }
      });
    case 'REMOVE_CLIENT':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};

export default clientsReducer;
