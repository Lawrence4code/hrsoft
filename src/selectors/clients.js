// get visible clients
import moment from 'moment';

const getVisibleClients = (clients, { input, sortBy, startDate, endDate }) => {
  // console.log(input);
  return clients
    .filter(client => {
      const createdAtMoment = moment(client.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;

      const firstNameMatch = client.firstName
        .toLowerCase()
        .includes(input.toLowerCase());
      const lastNameMatch = client.lastName
        .toLowerCase()
        .includes(input.toLowerCase());

      const phoneMatch = client.phone.includes(input);

      return (
        startDateMatch &&
        endDateMatch &&
        (firstNameMatch || lastNameMatch || phoneMatch)
      );
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'alphabetically') {
        return a.firstName < b.firstName ? -1 : 1;
      } else if (sortBy === 'lastCommunication') {
        return a.lastCommunication > b.lastCommunication ? 1 : -1;
      } else if (sortBy === 'priority') {
        return a.priority > b.priority ? -1 : 1;
      } else if (sortBy === 'memberType') {
        return a.memberType > b.memberType ? -1 : 1;
      }
    });
};

export default getVisibleClients;
