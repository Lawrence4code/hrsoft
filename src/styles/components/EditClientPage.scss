.editClientPage {
  &__content {
    height: 10rem;
    width: 90%;
    margin: 0 auto;
    display: grid;
    align-items: center;

    & h1 {
      color: $secondary-grey;
      font-weight: 300;
    }
  }

  &__heading {
    background-color: $light-back-grey;
  }
}
