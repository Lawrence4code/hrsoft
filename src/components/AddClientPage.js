import React from 'react';
import ClientForm from './ClientForm';
import { connect } from 'react-redux';
import { startAddClient } from '../actions/clients';

export class AddClientPage extends React.Component {
  onSubmit = client => {
    this.props.startAddClient(client);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="addClientPage">
        <div className="addClientPage__heading">
          <div className="addClientPage__content">
            <h1> Add New Client </h1>
          </div>
        </div>

        <ClientForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddClient: client => dispatch(startAddClient(client))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddClientPage);
