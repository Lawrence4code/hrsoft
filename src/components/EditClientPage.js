import React from 'react';
import ClientForm from './ClientForm';
import { connect } from 'react-redux';
import { startEditClient } from '../actions/clients';

export class EditClientPage extends React.Component {
  onSubmit = (client) => {
    console.log(this.props.client.id, '***', client);
    this.props.startEditClient(this.props.client.id, client);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="editClientPage">
        <div className="editClientPage__heading">
          <div className="editClientPage__content">
            <h1>Edit Client Details</h1>
          </div>
        </div>
        <ClientForm client={this.props.client} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    client: state.clients.find((client) => client.id === props.match.params.id),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startEditClient: (id, client) => {
      return dispatch(startEditClient(id, client));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditClientPage);
