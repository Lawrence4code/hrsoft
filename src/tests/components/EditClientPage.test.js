import React from 'react';
import { shallow } from 'enzyme';
import { EditClientPage } from '../../components/EditClientPage';
import clients from '../fixtures/clients';

let startEditClient, history, wrapper;

beforeEach(() => {
  startEditClient = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditClientPage
      startEditClient={startEditClient}
      history={history}
      client={clients[1]}
    />
  );
});

test('should render EditClientPage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render EditClientPage', () => {
  wrapper.find('ClientForm').prop('onSubmit')(clients[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditClient).toHaveBeenLastCalledWith(clients[1].id, clients[1]);
});
