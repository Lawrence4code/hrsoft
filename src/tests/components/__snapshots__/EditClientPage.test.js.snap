// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditClientPage 1`] = `
<div
  className="editClientPage"
>
  <div
    className="editClientPage__heading"
  >
    <div
      className="editClientPage__content"
    >
      <h1>
        Edit Client Details
      </h1>
    </div>
  </div>
  <ClientForm
    client={
      Object {
        "address": "FakeOne High Street",
        "createdAt": -345600000,
        "email": "Jane@doe.com",
        "field": "n/a",
        "firstName": "Jane",
        "gender": "F",
        "id": "2",
        "lastCommuniation": "",
        "lastName": "Doe",
        "memberType": "Silver",
        "note": "Not needed",
        "phone": "123456789",
        "status": "Housewife",
        "title": "Mrs",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
